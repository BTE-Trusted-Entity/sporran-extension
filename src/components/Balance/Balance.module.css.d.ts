export const __esModule: true;
export const balanceLine: string;
export const showBreakdown: string;
export const hideBreakdown: string;
export const breakdown: string;
export const balance: string;
export const connecting: string;
