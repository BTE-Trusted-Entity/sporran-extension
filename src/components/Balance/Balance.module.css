.balanceLine {
  composes: smallText from '../Typography/Typography.module.css';
  font-size: 1.25rem;
  margin: 1rem 0 0;
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
}

.showBreakdown {
  composes: buttonIcon from '../../components/Button/Button.module.css';
  background-image: url('../../components/images/chevron-down-blue.svg');
  width: 1.625rem;
  height: 1.625rem;
  background-size: 0.75rem;
  margin: 0 0 0 0.5rem;
}

.hideBreakdown {
  composes: showBreakdown;
  transform: rotate(180deg);
}

.breakdown {
  composes: smallText from '../../components/Typography/Typography.module.css';
  list-style: none;
  padding: 0;
  margin: 0.25rem 0 0;
}

.balance {
  margin: 0.25rem 0;
}

.connecting {
  composes: spinner from '../../components/LoadingSpinner/LoadingSpinner.module.css';
  position: absolute;
  left: 50%;
  transform: translateX(-50%);
}
