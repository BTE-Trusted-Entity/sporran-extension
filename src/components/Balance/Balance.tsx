import { Fragment, useEffect, useState } from 'react';
import BN from 'bn.js';
import { browser } from 'webextension-polyfill-ts';

import * as styles from './Balance.module.css';

import { KiltAmount } from '../KiltAmount/KiltAmount';

import {
  BalanceChange,
  onAddressBalanceChange,
} from '../../utilities/balanceChanges/balanceChanges';
import { useBooleanState } from '../../utilities/useBooleanState/useBooleanState';

interface BalanceBN {
  transferable: BN;
  usableForFees: BN;
  bonded: BN;
  locked: BN;
  total: BN;
}

export function useAddressBalance(address: string): BalanceBN | null {
  const [balance, setBalance] = useState<BalanceBN | null>(null);

  useEffect(() => {
    setBalance(null);

    return onAddressBalanceChange(address, (error, output?: BalanceChange) => {
      if (error || !output) {
        console.error(error);
      } else {
        if (output.address === address) {
          setBalance(output.balances as BalanceBN);
        }
      }
    });
  }, [address]);

  return balance;
}

interface BalanceProps {
  address: string;
  breakdown?: boolean;
  smallDecimals?: boolean;
}

export function Balance({
  address,
  breakdown = false,
  smallDecimals = false,
}: BalanceProps): JSX.Element {
  const t = browser.i18n.getMessage;
  const balance = useAddressBalance(address);
  const connecting = balance === null;

  const breakdownVisibility = useBooleanState();

  return (
    <>
      <p className={styles.balanceLine}>
        {connecting && <span className={styles.connecting}></span>}
        {t('component_Balance_label')}
        {!connecting && (
          <KiltAmount
            amount={balance.total}
            type="funds"
            smallDecimals={smallDecimals}
          />
        )}

        {breakdown &&
          !connecting &&
          (breakdownVisibility.current ? (
            <button
              type="button"
              onClick={breakdownVisibility.off}
              className={styles.hideBreakdown}
              title={t('component_Balance_hideBreakdown')}
              aria-label={t('component_Balance_hideBreakdown')}
            />
          ) : (
            <button
              type="button"
              onClick={breakdownVisibility.on}
              className={styles.showBreakdown}
              title={t('component_Balance_showBreakdown')}
              aria-label={t('component_Balance_showBreakdown')}
            />
          ))}
      </p>
      {breakdownVisibility.current && !connecting && (
        <Fragment>
          <ul className={styles.breakdown}>
            <li className={styles.balance}>
              {t('component_Balance_transferable')}
              <KiltAmount amount={balance.transferable} type="funds" />
            </li>
            <li className={styles.balance}>
              {t('component_Balance_locked')}
              <KiltAmount amount={balance.locked} type="funds" />
            </li>
            <li className={styles.balance}>
              {t('component_Balance_bonded')}
              <KiltAmount amount={balance.bonded} type="funds" />
            </li>
          </ul>
        </Fragment>
      )}
    </>
  );
}
