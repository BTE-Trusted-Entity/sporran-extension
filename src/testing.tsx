export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
