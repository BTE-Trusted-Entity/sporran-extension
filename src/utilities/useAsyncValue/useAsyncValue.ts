import { useEffect, useState } from 'react';

import { exceptionToError } from '../exceptionToError/exceptionToError';

export function useAsyncValue<Input extends Array<unknown>, Output>(
  fetcher: (...args: Input) => Promise<Output>,
  args: Input,
): Output | undefined {
  const [data, setData] = useState<Output>();
  const [error, setError] = useState<Error>();

  useEffect(() => {
    (async () => {
      try {
        setData(await fetcher(...args));
      } catch (exception) {
        console.error(exception);
        setError(exceptionToError(exception));
      }
    })();
  }, args); // eslint-disable-line react-hooks/exhaustive-deps

  if (error) {
    throw error;
  }

  return data;
}
