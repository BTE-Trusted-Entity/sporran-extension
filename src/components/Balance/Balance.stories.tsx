import { Meta } from '@storybook/react';

import { Balance } from './Balance';

export default {
  title: 'Components/Balance',
  component: Balance,
} as Meta;

export function Template(): JSX.Element {
  return <Balance address="4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1" />;
}

export function WithBreakdown(): JSX.Element {
  return (
    <Balance
      address="4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
      breakdown
      smallDecimals
    />
  );
}
