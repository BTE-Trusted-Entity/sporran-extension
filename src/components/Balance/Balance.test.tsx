import userEvent from '@testing-library/user-event';

import { render, screen } from '../../testing/testing';
import { useAsyncValue } from '../../utilities/useAsyncValue/useAsyncValue';

import { Balance } from './Balance';

const mockAddress = '4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1';

jest.mock('../../utilities/useAsyncValue/useAsyncValue');
jest.mocked(useAsyncValue).mockReturnValue(false);

describe('Balance', () => {
  it('should render', async () => {
    const { container } = render(<Balance address={mockAddress} />);

    await screen.findByLabelText('KILT Coin');

    expect(container).toMatchSnapshot();
  });

  it('should show balance breakdown with update balance button', async () => {
    const { container } = render(
      <Balance address={mockAddress} breakdown smallDecimals />,
    );

    const showBreakdown = await screen.findByRole('button', {
      name: 'Show balance breakdown',
    });
    await userEvent.click(showBreakdown);

    expect(container).toMatchSnapshot();
  });
});
