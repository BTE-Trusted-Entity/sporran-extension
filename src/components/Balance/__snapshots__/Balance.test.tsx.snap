// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Balance should render 1`] = `
<div>
  <p
    class="balanceLine"
  >
    Total Balance: 
    <span
      class="amount"
    >
      1.2340
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
    </span>
  </p>
</div>
`;

exports[`Balance should show balance breakdown with update balance button 1`] = `
<div>
  <p
    class="balanceLine"
  >
    Total Balance: 
    <span
      class="amount"
    >
      1
      <span
        class="fractional"
      >
        .2340
      </span>
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
    </span>
    <button
      aria-label="Hide balance breakdown"
      class="hideBreakdown"
      title="Hide balance breakdown"
      type="button"
    />
  </p>
  <ul
    class="breakdown"
  >
    <li
      class="balance"
    >
      Transferable Balance: 
      <span
        class="amount"
      >
        1.2160
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </li>
    <li
      class="balance"
    >
      Locked Balance: 
      <span
        class="amount"
      >
        0.0100
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </li>
    <li
      class="balance"
    >
      Bonded Balance: 
      <span
        class="amount"
      >
        0.0080
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </li>
  </ul>
</div>
`;
